src/Microsoft.Identity.Client/Azure/Exponential.cs

```diff
@@ -0,0 +1,131 @@
+using System;
+using System.Diagnostics;
+using System.Net.Http;
+using System.Threading.Tasks;
+
+namespace Microsoft.Identity.Client.Azure
+{
+    /// <summary>
+    /// RetryWithExponentialBackoff runs a task with an exponential backoff
+    /// </summary>
+    public sealed class RetryWithExponentialBackoff
+    {
+        private readonly int _maxRetries, _delayMilliseconds, _maxDelayMilliseconds;
+
+        /// <summary>
+        /// Create an instance of a RetryWithExponentialBackoff
+        /// </summary>
+        /// <param name="maxRetries">maximum number of retries</param>
+        /// <param name="delayMilliseconds">initial delay in milliseconds</param>
+        /// <param name="maxDelayMilliseconds">maximum delay in milliseconds</param>
+        public RetryWithExponentialBackoff(int maxRetries = 50, int delayMilliseconds = 100, int maxDelayMilliseconds = 2000)
+        {
+            _maxRetries = maxRetries;
+            _delayMilliseconds = delayMilliseconds;
+            _maxDelayMilliseconds = maxDelayMilliseconds;
+        }
+
+        /// <summary>
+        /// RunAsync will attempt to execute the a task with a exponential retry
+        /// </summary>
+        /// <param name="func">task to execute</param>
+        /// <returns>exponentially backed off task</returns>
+        public async Task RunAsync(Func<Task> func)
+        {
+            var backoff = new ExponentialBackoff(_maxRetries, _delayMilliseconds, _maxDelayMilliseconds);
+        retry:
+            try
+            {
+                await func().ConfigureAwait(false);
+            }
+            catch (Exception ex) when (ex is TimeoutException || ex is HttpRequestException || ex is TaskCanceledException || ex is TransientManagedIdentityException)
+            {
+                var msg = $"Exception raised is: {ex.GetType()} Message: {ex.Message} -- Inner Message: {ex.InnerException?.Message}";
+                Debug.WriteLine(msg);
```

Comment:
Why not just `Debug.WriteLine(ex.ToString())` to expose the stack as well as the message? Recursion through InnerException if needed as well

Reply:
Good idea. I think I will also need to pipe in the logger implementation from MSAL, but not quite clear on it. @DAllanCarr, what do you think?

Reply:
I'm not sure we need to log but cool with the depenepdency . Obviously surfacing the final exception is more important.